### PR TITLE
[IMP] payment: add distinction between providers and tokens

### PR DESCRIPTION
**Pull request by anko-odoo**
Currently there is no distinction between saved tokens and providers and there is no info on tokens under what provider they were created. When client has more than one token it looks messy and counters the point of token existance. Now tokens have their creation date,provider and are no longer in the same card as providers.

task-2510973

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/payment/views/payment_templates.xml [hunk 1]
```diff
@@ -315,6 +342,7 @@
         </form>
     </template>
 
+
```
Comment (Demesmaeker) [inline addons/payment/views/payment_templates.xml:345]: Why? ^^

File: addons/payment/views/payment_templates.xml [hunk 2]
```diff
@@ -411,7 +439,7 @@
 
     <!-- Test token badge -->
     <template id="test_token_badge" name="Test Token Badge">
-        <span t-if="provider.state == 'test'" class="badge rounded-pill test-bg-warning">
+        <span t-if="token.provider_id.state == 'test'" class="badge rounded-pill text-bg-warning">
```
Comment (Demesmaeker) [inline addons/payment/views/payment_templates.xml:442]: Good catch!

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello there,
I tried to test this task on the runbot, but I immediately got a traceback due to `website_sale_picking` inheriting `checkout_payment` here: https://github.com/odoo/odoo/blob/00b1edbb2547db397841738968a8cb6c50ae7ca8/addons/website_sale_picking/views/templates.xml#L9-L13
So it attached itself to the token part of your code, and raised the `KeyError: 'provider'`

File: addons/payment/views/payment_templates.xml (hunk 1 above)
Comment (anko-odoo) [reply addons/payment/views/payment_templates.xml:345]: That is good question, must have slipped my mind.

File: addons/payment/views/payment_templates.xml [hunk 3]
```diff
@@ -209,104 +223,118 @@
...
+                            <!-- === "Delete" token button === -->
+                            <button name="o_payment_delete_token"
+                                    class="btn btn-primary btn-sm float-end">
+                                <i class="fa fa-trash"/> Delete
+                            </button>
+                            <div>
+                                <span class="fst-italic small"> 
+                                    Managed by
+                                    <span class="token_payment_provider_name" t-esc="token.provider_id.name"/>
+                                    - Created on
```
Comment (Demesmaeker) [inline addons/payment/views/payment_templates.xml:263]:
```suggestion
                                    - created on 
```

As per the Spec of the task :shrug:

File: addons/payment/views/payment_templates.xml [hunk 4]
```diff
@@ -209,104 +223,118 @@
...
+                                    Unpublished
+                                </span>
+                            </label>
+                            <!-- === "Delete" token button === -->
+                            <button name="o_payment_delete_token"
+                                    class="btn btn-primary btn-sm float-end">
+                                <i class="fa fa-trash"/> Delete
+                            </button>
+                            <div>
+                                <span class="fst-italic small">
```
Comment (Demesmaeker) [inline addons/payment/views/payment_templates.xml:260]: It'll be better with:
```suggestion
                                <span class="fst-italic small text-muted"> 
```

**Review by Demesmaeker — COMMENTED** [review]
Hello ^^

File: addons/payment/views/payment_templates.xml [hunk 5]
```diff
@@ -56,14 +56,72 @@
...
+                                        Fees
+                                </span>
+                                <!-- === "Test token" badge === -->
+                                <t t-call="payment.test_token_badge"/>
+                                <!-- === "Unpublished" badge === -->
+                                <span t-if="not token.provider_id.is_published" class="badge rounded-pill text-bg-danger ms-1">
+                                    Unpublished
+                                </span>
+                                <div class="ms-4">
+                                    <span class="fst-italic small ">
```
Comment (Demesmaeker) [inline addons/payment/views/payment_templates.xml:91]: Here too =]
```suggestion
                                    <span class="fst-italic small text-muted">
```

**Review by Demesmaeker — APPROVED** [review]

File: addons/payment/views/payment_templates.xml [hunk 6]
```diff
@@ -56,14 +56,72 @@
...
+                    <t t-foreach="tokens" t-as="token">
+                        <div name="o_payment_option_card" class="card-body o_payment_option_card">
+                            <label>
+                                <!-- === Radio button === -->
+                                <input name="o_payment_radio"
+                                    type="radio"
+                                    t-att-checked="token.id == default_payment_option_id"
+                                    t-att-data-payment-option-id="token.id"
+                                    t-att-data-provider="token.provider_code"
+                                    data-payment-option-type="token"/>
```
Comment (Demesmaeker) [inline addons/payment/views/payment_templates.xml:72]: Could you please realign all the attributes as that as the case before on the whole diff? ^^

File: addons/payment/views/payment_templates.xml [hunk 7]
```diff
@@ -56,14 +56,72 @@
...
+                                <!-- === "Test token" badge === -->
+                                <t t-call="payment.test_token_badge"/>
+                                <!-- === "Unpublished" badge === -->
+                                <span t-if="not token.provider_id.is_published" class="badge rounded-pill text-bg-danger ms-1">
+                                    Unpublished
+                                </span>
+                                <div class="ms-4">
+                                    <span class="fst-italic small text-muted">
+                                    Managed by
+                                    <span class="token_payment_provider_name" t-esc="token.provider_id.name"/>
```
Comment (Demesmaeker) [inline addons/payment/views/payment_templates.xml:93]: I saw that the spec ask for the provider code here, instead of the name. Why did you choose to go with the name nonetheless?
Since I also think the name is slightly better, even if it has important downsides, I'll approve the PR. This could be an interesting point to discuss with @dalaOdoo though

**Review by Demesmaeker — APPROVED** [review]
As these comments are not blocking, I'll green the PR, so it could move, but ask dala to have confirmation for the name/code =]

File: addons/payment/views/payment_templates.xml (hunk 7 above)
Comment (anko-odoo) [reply addons/payment/views/payment_templates.xml:93]: I don't know if it's a good enough argument, but "code" is starting from lowercase. Also, another case is that if someone would duplicate the provider and give them another name if we would use code instead of name it would still show the same code, which might be a desirable thing depending on what we want actually. But I used name since in the specs it was uppercase.

Comment (Demesmaeker) [reply addons/payment/views/payment_templates.xml:93]: More important: we have the `payment_custom` module. I know, there isn't any token created by this provider, so this is irrelevant, but this is the main argument on why using the name instead of the code makes sense in my opinion. But then, relying on the code instead of the name makes sure the displayed provider name is coherent and not `Stripe Main Account for Company Thingy thing`, as that could be the case if the provider was renamed.

tldr; using the code is making sure the displayed name is correct and not personalized.

File: addons/payment/views/payment_templates.xml [hunk 8]
```diff
@@ -56,14 +56,79 @@
             <t t-set="footer_template_id"
                t-value="footer_template_id or 'payment.footer'"/>
 
+            <!-- === Tokens === -->
+            <t t-if="token_count>0">
+                <div class="mt-3 fs-4">Saved payment methods</div>
+                <div class="card">
+                    <t t-foreach="tokens" t-as="token">
+                        <t t-set="provider" t-value="token.provider_id"/>
```
Comment (AntoineVDV) [inline addons/payment/views/payment_templates.xml:64]: It's a good idea to extract `token.provider_id` in a variable, but you should use it more. There are several occurrences of `token.provider_id` that still coexist with `provider`, where I'd expect to find only `provider`.

File: addons/payment/views/payment_templates.xml [hunk 9]
```diff
@@ -56,14 +56,79 @@
...
+                                <span t-if="not token.provider_id.is_published" class="badge rounded-pill text-bg-danger ms-1">
+                                    Unpublished
+                                </span>
+                                <div class="ms-4">
+                                    <span class="fst-italic small text-muted">
+                                        Managed by
+                                        <!-- It would not work with custom payment provider e.g. 
+                                          "Wire transfer" as it would just show "custom", but
+                                          for now we don't tokenize custom providers, so it's
+                                          not a problem -->
```
Comment (AntoineVDV) [inline addons/payment/views/payment_templates.xml:97]: Good initiative to write your findings down, but it creates more noise than it is useful, IMHO. If someone ever adds tokenization for a custom provider, they will find out that it doesn't work before reading this comment, and they will easily fix the issue anyway.

File: addons/payment/views/payment_templates.xml [hunk 10]
```diff
@@ -56,14 +56,79 @@
...
+                                        Fees
+                                </span>
+                                <!-- === "Test token" badge === -->
+                                <t t-call="payment.test_token_badge"/>
+                                <!-- === "Unpublished" badge === -->
+                                <span t-if="not token.provider_id.is_published" class="badge rounded-pill text-bg-danger ms-1">
+                                    Unpublished
+                                </span>
+                                <div class="ms-4">
+                                    <span class="fst-italic small text-muted">
```
Comment (AntoineVDV) [inline addons/payment/views/payment_templates.xml:92]: These two could be merged together. Also, it's better not to use `span` for managing block structure.

File: addons/payment/views/payment_templates.xml [hunk 11]
```diff
@@ -56,14 +56,79 @@
...
+                                        Managed by
+                                        <!-- It would not work with custom payment provider e.g. 
+                                          "Wire transfer" as it would just show "custom", but
+                                          for now we don't tokenize custom providers, so it's
+                                          not a problem -->
+                                        <span class="token_payment_provider_code"
+                                              t-esc="dict(provider._fields['code']._description_selection(provider.env))[provider.code]"/>
+                                        - created on
```
Comment (AntoineVDV) [inline addons/payment/views/payment_templates.xml:100]: Personal opinion: I think it looks better with an *em* dash.
```suggestion
                                        — created on 
```

File: addons/payment/views/payment_templates.xml [hunk 12]
```diff
@@ -56,14 +56,79 @@
...
+                                    <span class="fst-italic small text-muted">
+                                        Managed by
+                                        <!-- It would not work with custom payment provider e.g. 
+                                          "Wire transfer" as it would just show "custom", but
+                                          for now we don't tokenize custom providers, so it's
+                                          not a problem -->
+                                        <span class="token_payment_provider_code"
+                                              t-esc="dict(provider._fields['code']._description_selection(provider.env))[provider.code]"/>
+                                        - created on 
+                                        <span class="token_payment_creation_date" t-esc="token.create_date.date()"/>
```
Comment (AntoineVDV) [inline addons/payment/views/payment_templates.xml:101]:
```suggestion
                                        <span class="token_payment_creation_date"
                                                    t-esc="token.create_date.date()"/>  
```

Comment (AntoineVDV) [reply addons/payment/views/payment_templates.xml:101]: Also, swap `class` and `t-esc`; the reader wants to know what is printed before how it's stylized.

File: addons/payment/views/payment_templates.xml [hunk 13]
```diff
@@ -56,14 +56,79 @@
...
+                                <t t-call="{{token_inline_form_xml_id}}">
+                                    <t t-set="token" t-value="token"/>
+                                </t>
+                            </div>
+                        </div>
+                    </t>
+                </div>
+            </t>
+            <!-- === Providers === -->
+            <div class="mt-3 fs-4" t-if="token_count>0">
```
Comment (AntoineVDV) [inline addons/payment/views/payment_templates.xml:123]: Swap the `class` and the `t-if`; the code structure comes before the element style. Also, some previous comments apply here too.

File: addons/payment/views/payment_templates.xml [hunk 14]
```diff
@@ -411,7 +452,7 @@
 
     <!-- Test token badge -->
     <template id="test_token_badge" name="Test Token Badge">
-        <span t-if="provider.state == 'test'" class="badge rounded-pill test-bg-warning">
+        <span t-if="token.provider_id.state == 'test'" class="badge rounded-pill text-bg-warning">
```
Comment (AntoineVDV) [inline addons/payment/views/payment_templates.xml:455]: Nice catch for the typo, but why do you change `provider` to `token.provider_id`?

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

**Comment by AntoineVDV** [comment]
@robodoo r-

**Comment by AntoineVDV** [comment]
@robodoo r+
